domain/externalcontroller/state/state.go

```diff
@@ -168,51 +166,83 @@ func (st *State) ImportExternalControllers(ctx context.Context, infos []crossmod
 
 func (st *State) updateExternalControllerTx(
 	ctx context.Context,
-	tx *sql.Tx,
+	tx *sqlair.TX,
 	ci crossmodel.ControllerInfo,
 ) error {
 	cID := ci.ControllerTag.Id()
-	q := `
+	updateControllerQueryString := `
 INSERT INTO external_controller (uuid, alias, ca_cert)
-VALUES (?, ?, ?)
-  ON CONFLICT(uuid) DO UPDATE SET alias=excluded.alias, ca_cert=excluded.ca_cert`
+VALUES ($ExternalController.uuid, $ExternalController.alias, $ExternalController.ca_cert)
+  ON CONFLICT(uuid) DO UPDATE SET alias=excluded.alias, ca_cert=excluded.ca_cert
+`
+	updateControllerStmt, err := sqlair.Prepare(updateControllerQueryString, ExternalController{})
+	if err != nil {
+		return errors.Trace(err)
+	}
 
-	if _, err := tx.ExecContext(ctx, q, cID, ci.Alias, ci.CACert); err != nil {
+	externalController := ExternalController{
+		ID:     cID,
+		Alias:  sql.NullString{String: ci.Alias, Valid: true},
+		CACert: ci.CACert,
+	}
+	if err := tx.Query(ctx, updateControllerStmt, externalController).Run(); err != nil {
 		return errors.Trace(err)
 	}
 
-	addrsBinds, addrsAnyVals := database.SliceToPlaceholder(ci.Addrs)
-	q = fmt.Sprintf(`
+	deleteQueryString := `
 DELETE FROM external_controller_address
-WHERE  controller_uuid = ?
-AND    address NOT IN (%s)`[1:], addrsBinds)
-
-	args := append([]any{cID}, addrsAnyVals...)
-	if _, err := tx.ExecContext(ctx, q, args...); err != nil {
+WHERE  controller_uuid = $ExternalController.uuid
+AND    address NOT IN ($uuids[:])
+`
+	deleteStmt, err := sqlair.Prepare(deleteQueryString, ExternalController{}, uuids{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+	cIDs := uuids(ci.Addrs)
+	if len(cIDs) == 0 {
+		cIDs = uuids{}
+	}
+	if err := tx.Query(ctx, deleteStmt, externalController, cIDs).Run(); err != nil {
 		return errors.Trace(err)
 	}
 
-	for _, addr := range ci.Addrs {
-		q := `
+	insertAddressQueryString := `
 INSERT INTO external_controller_address (uuid, controller_uuid, address)
-VALUES (?, ?, ?)
-  ON CONFLICT(controller_uuid, address) DO NOTHING`
+VALUES ($ExternalControllerAddress.uuid, $ExternalControllerAddress.controller_uuid,
```

Comment:
It is very important to do the least work we can inside a write transaction, as there can only be one at a time.
That means we don't want to be preparing if we can avoid it.

Move the statement preparations into another method that is called in `UpdateExternalController` and `ImportExternalControllers`, and pass the returned statements to this method for execution.

Reply:
I've moved the preparation to a separate function `prepareUpdateStatements`.

I've put the struct to contain the prepared statements in `types.go` and annotated it with descriptions of the queries. I was wondering, should this struct should live there, or be beside the `prepareUpdateStatements` function? Right now it is quite far from the queries themselves.